File: src/shared/api/bookmarks.ts
import { invoke } from '@tauri-apps/api/core';
import type { ReadingPosition } from './reader';
import { TextRange } from '../types/note';

// Types matching backend structs
export type BookmarkKind = 'regular' | 'custom';

export type Bookmark = {
  id: string;
  book_path: string;
  position: ReadingPosition;
  preview: string;
  kind: BookmarkKind;
  created_at: number;
  range: TextRange;
};

// API functions
export async function addBookmark(
	bookPath: string,
	position: ReadingPosition,
	range: TextRange,
	preview: string,
	kind: BookmarkKind = 'regular',
): Promise<Bookmark> {
	return invoke<Bookmark>('add_bookmark', { bookPath, position, range, preview, kind });
}

export async function getBookmarks(bookPath?: string): Promise<Bookmark[]> {
  return invoke<Bookmark[]>('get_bookmarks', { bookPath: bookPath ?? null });
}

export async function deleteBookmark(bookmarkId: string): Promise<void> {
  return invoke('delete_bookmark', { bookmarkId });
}

export async function getBookmark(bookmarkId: string): Promise<Bookmark | null> {
  return invoke<Bookmark>('get_bookmark', { bookmarkId });
}

File: src-tauri/src/core/reader/bookmarks.rs
use rayon::range;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::core::reader::{ReadingPosition, TextRange};

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub enum BookmarkKind {
    Regular,
    Custom,
}

impl Default for BookmarkKind {
    fn default() -> Self {
        Self::Regular
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Bookmark {
    pub id: String,
    pub book_path: String,
    pub position: ReadingPosition,
    pub preview: String,
    pub kind: BookmarkKind,
    pub created_at: i64,
    pub range: TextRange
}

impl Bookmark {
    pub fn new(
        book_path: String,
        position: ReadingPosition,
        preview: String,
        kind: BookmarkKind,
        created_at: i64,
        range: TextRange,
    ) -> Self {
        Self {
            id: Uuid::new_v4().to_string(),
            book_path,
            position,
            preview,
            kind,
            created_at,
            range
        }
    }
}
